pkg/operator/resource/resourceapply/monitoring.go

```diff
@@ -2,148 +2,168 @@ package resourceapply
+}
+
+// DeleteServiceMonitor deletes the ServiceMonitor.
+func DeleteServiceMonitor(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
+	return DeleteUnstructuredResource(ctx, client, recorder, required, serviceMonitorGVR)
+}
+
+// ApplyUnstructuredResourceImproved can utilize the cache to reconcile the existing resource to the desired state.
+// NOTE: A `nil` defaultingFunc and equalityChecker are assigned resourceapply.noDefaulting and equality.Semantic,
+// respectively. Users are recommended to instantiate a cache to benefit from the memoization machinery.
+func ApplyUnstructuredResourceImproved(
+	ctx context.Context,
+	client dynamic.Interface,
+	recorder events.Recorder,
+	required *unstructured.Unstructured,
+	cache ResourceCache,
+	resourceGVR schema.GroupVersionResource,
+	defaultingFunc mimicDefaultingFunc,
+	equalityChecker equalityChecker,
+) (*unstructured.Unstructured, bool, error) {
+	name := required.GetName()
 	namespace := required.GetNamespace()
-	existing, err := client.Resource(serviceMonitorGVR).Namespace(namespace).Get(ctx, required.GetName(), metav1.GetOptions{})
+
+	// Create if resource does not exist, and update cache with new metadata.
+	if cache == nil {
+		cache = noCache
+	}
+	existing, err := client.Resource(resourceGVR).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
-		newObj, createErr := client.Resource(serviceMonitorGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
-		if createErr != nil {
-			recorder.Warningf("ServiceMonitorCreateFailed", "Failed to create ServiceMonitor.monitoring.coreos.com/v1: %v", createErr)
-			return nil, true, createErr
-		}
-		recorder.Eventf("ServiceMonitorCreated", "Created ServiceMonitor.monitoring.coreos.com/v1 because it was missing")
-		return newObj, true, nil
+		want, err := client.Resource(resourceGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
+		reportCreateEvent(recorder, required, err)
+		cache.UpdateCachedResourceMetadata(required, want)
+		return want, true, err
 	}
 	if err != nil {
 		return nil, false, err
 	}
 
-	existingCopy := existing.DeepCopy()
+	// Skip if the cache is non-nil, and the metadata hashes and resource version hashes match.
+	if cache.SafeToSkipApply(required, existing) {
+		return existing, false, nil
+	}
 
-	toUpdate, modified, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
+	// Ensure metadata field is present on the object.
+	existingCopy := existing.DeepCopy()
+	existingObjectMeta, found, err := unstructured.NestedMap(existingCopy.Object, "metadata")
 	if err != nil {
 		return nil, false, err
 	}
-
-	if !modified {
-		return nil, false, nil
+	if !found {
+		return nil, false, errorsstdlib.New(fmt.Sprintf("metadata not found in %s", existingCopy.GetName()))
```

Comment:
Should these error returns maybe contain context about the object where the error occurred? Here its the existing object, further down the required?

Reply:
Hmm, so `found` here corresponds to the [`existingObject`](https://github.com/openshift/library-go/pull/1575/files#diff-55de3da9293d86816615a21ec35c9725ab821c04b938de22d815679554b2fe92R105) and the [`requiredObject`](https://github.com/openshift/library-go/pull/1575/files#diff-55de3da9293d86816615a21ec35c9725ab821c04b938de22d815679554b2fe92R112) below. Do you mean appending something more than just the object name in the error messages? LMK if I missed something.

Reply:
Sure I understand when the error is returned. But iiuc the error returned will always be something like `metadata not found in prometheus-k8s`, whether its the existing copy or the required object. Adding information which object generates an error might be a good idea?

Reply:
Oh, you mean the object key (as in `namespace/object`) for a better context? Done!